### PR TITLE
Support using remote HDFS as storage in Raptor

### PR DESCRIPTION
**Pull request by jessesleeping**
TBD

```
== RELEASE NOTES ==

Raptor Changes
* Change `storage.data-directory` from path to URI. For existing deployment on local flash, a scheme header "file://" should be added to the original config value.
* Change error code name `RAPTOR_LOCAL_FILE_SYSTEM_ERROR` to `RAPTOR_FILE_SYSTEM_ERROR`.
```

### CODE REVIEW

File: presto-raptor/src/main/java/com/facebook/presto/raptor/filesystem/RaptorRemoteFileSystemConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.raptor.filesystem;
+
+import org.apache.hadoop.conf.Configuration;
+
+public interface RaptorRemoteFileSystemConfiguration
+{
+    // TODO: pass HdfsContext to this method
```
Comment (highker) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/filesystem/RaptorRemoteFileSystemConfiguration.java]: Let's pass `HdfsContext` to `getConfiguration()`

File: presto-raptor/src/main/java/com/facebook/presto/raptor/filesystem/HdfsOrcDataEnvironment.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+        try {
+            this.fileSystem = baseLocation.getFileSystem(configuration.getConfiguration());
+        }
+        catch (IOException e) {
+            throw new PrestoException(RAPTOR_FILE_SYSTEM_ERROR, "Raptor cannot create HDFS file system", e);
+        }
+    }
+
+    @Override
+    public FileSystem getFileSystem()
```
Comment (highker) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/filesystem/HdfsOrcDataEnvironment.java]: Create another interface for `OrcDataEnvironment` so we can pass the query info and log it:
```java
public FileSystem getFileSystem(HdfsContext context)
```
The implementation of it for this class could be
```java
public FileSystem getFileSystem(HdfsContext context)
{
    return baseLocation.getFileSystem(configuration.getConfiguration(context));
}

File: presto-raptor/src/main/java/com/facebook/presto/raptor/filesystem/FileSystemModule.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+                .put("local", new LocalFileSystemModule())
+                .put("hdfs", new HdfsModule())
+                .putAll(providers)
+                .build();
+    }
+
+    @Override
+    protected void setup(Binder binder)
+    {
+        configBinder(binder).bindConfig(FileSystemConfig.class);
```
Comment (highker) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/filesystem/FileSystemModule.java]: Take `StorageManagerConfig.class`

File: presto-raptor/src/main/java/com/facebook/presto/raptor/storage/StorageModule.java [hunk 4]
```diff
@@ -31,16 +31,16 @@
...
 import com.google.inject.Scopes;
+import io.airlift.configuration.AbstractConfigurationAwareModule;
 
 import static io.airlift.configuration.ConfigBinder.configBinder;
 import static java.util.Objects.requireNonNull;
 import static org.weakref.jmx.ObjectNames.generatedNameOf;
 import static org.weakref.jmx.guice.ExportBinder.newExporter;
 
 public class StorageModule
-        implements Module
```
Comment (highker) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/storage/StorageModule.java]: No need to change

File: presto-raptor/src/main/java/com/facebook/presto/raptor/filesystem/FileStorageService.java [hunk 5]
```diff
@@ -55,9 +59,9 @@
     private final File baseQuarantineDir;
 
     @Inject
-    public FileStorageService(OrcDataEnvironment environment, StorageManagerConfig config)
+    public FileStorageService(OrcDataEnvironment environment, StorageManagerConfig storageManagerConfig, RaptorLocalFileSystemConfig localFileSystemConfig)
```
Comment (highker) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/filesystem/FileStorageService.java]: Call it `LocalFileStorageService`

File: presto-raptor/src/main/java/com/facebook/presto/raptor/filesystem/HdfsOrcDataEnvironment.java [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+import javax.inject.Inject;
+
+import java.io.IOException;
+
+import static com.facebook.presto.raptor.RaptorErrorCode.RAPTOR_FILE_SYSTEM_ERROR;
+
+public class HdfsOrcDataEnvironment
+        implements OrcDataEnvironment
+{
+    private final FileSystem fileSystem;
```
Comment (highker) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/filesystem/HdfsOrcDataEnvironment.java]: remove this

File: presto-raptor/src/main/java/com/facebook/presto/raptor/filesystem/HdfsStorageService.java [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+
+    private final FileSystem fileSystem;
+
+    private final Path baseStorageDir;
+    private final Path baseQuarantineDir;
+
+    @Inject
+    public HdfsStorageService(OrcDataEnvironment environment, Path baseLocation)
+    {
+        this.fileSystem = requireNonNull(environment, "requireNonNull is null").getFileSystem();
```
Comment (highker) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/filesystem/HdfsStorageService.java]:
```java
this.fileSystem = requireNonNull(environment, "requireNonNull is null").getFileSystem(EMPTY_CONTEXT, baseLocation);

File: presto-raptor/src/main/java/com/facebook/presto/raptor/storage/StorageManagerConfig.java [hunk 8]
```diff
@@ -72,12 +72,13 @@
     private String shardDayBoundaryTimeZone = TimeZoneKey.UTC_KEY.getId();
     private int maxAllowedFilesPerWriter = Integer.MAX_VALUE;
 
-    @NotNull
+    @Deprecated
     public File getDataDirectory()
```
[the remark this answers is not in the document]
Comment (jessesleeping) [reply presto-raptor/src/main/java/com/facebook/presto/raptor/storage/StorageManagerConfig.java]: I am thinking using URI scheme as "FileSystemModule" selector is not very feasible. For example, under "HdfsStorageModule" mode, we can support URI schemes for both "hdfs" and "local". It's a many to many mapping so I suggest we still use a separate config to indicate what file system we are going to use.

**Pull request merged**